### PR TITLE
[Fixes #3016] Disable response buffering in places where the content is already built/available

### LINKED ISSUE #3016: Disable response buffering in places where the content is already built/available.
We should check for the presence of `IHttpBufferingFeature` on the context and if present should disable response buffering at least in the following cases:

- https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Core/FileStreamResult.cs#L77
- https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Core/StreamOutputFormatter.cs#L43
- https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Core/ContentResult.cs#L70
- https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Core/FileContentResult.cs#L69

@Tratcher @rynowak

### PR DESCRIPTION
**Pull request by kichalla**
@rynowak @Tratcher

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ContentResult.cs [hunk 1]
```diff
@@ -54,6 +55,12 @@ public class ContentResult : ActionResult
 
             if (Content != null)
             {
+                var bufferingFeature = response.HttpContext.GetFeature<IHttpBufferingFeature>();
+                if (bufferingFeature != null)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/ContentResult.cs]: `bufferingFeature?.DisableResponseBuffering();`

**Comment by pranavkm** [comment]
Could this also apply to Razor views?

File: src/Microsoft.AspNet.Mvc.Core/ContentResult.cs (hunk 1 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/ContentResult.cs]: :+1:

**Comment by kichalla** [comment]
> Could this also apply to Razor views?

Currently I am disabling buffering only for scenarios where the content is already available. For  razor views or formatter serialization the final content is not available yet and the scenario is that one might want to enable buffering to catch any exceptions that happen during the generation of the final content in order to send back a friendly error message instead of partial content (just fyi...this is similar to web api 2 behavior).

File: src/Microsoft.AspNet.Mvc.Core/ContentResult.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ContentResult.cs]: PranavPoints++

File: test/Microsoft.AspNet.Mvc.Core.Test/ContentResultTest.cs [hunk 2]
```diff
@@ -183,9 +213,33 @@ private static ActionContext GetActionContext(HttpContext httpContext)
...
+
+            if (addBufferingFeature)
+            {
+                context.SetFeature<IHttpBufferingFeature>(new TestBufferingFeature());
+            }
+
+            return context;
+        }
+
+        private class TestBufferingFeature : IHttpBufferingFeature
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ContentResultTest.cs]: IMO just use Moq or lift this into a top-level class so you can reuse it.

**Comment by kichalla** [comment]
Updated. cc: @pranavkm @rynowak

File: test/Microsoft.AspNet.Mvc.Core.Test/ContentResultTest.cs [hunk 3]
```diff
@@ -78,6 +80,34 @@ public class ContentResultTest
             Assert.Equal("text/plain; charset=utf-7", httpContext.Response.ContentType);
         }
 
+        [Fact]
+        public async Task ContentResult_DisablesResponseBuffering_IfBufferingFeatureAvailable()
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/ContentResultTest.cs]: Is there a complementary scenario `IfBufferingFeatureIsNotAvailable` already covered? If not, could we add one?

File: test/Microsoft.AspNet.Mvc.Core.Test/ContentResultTest.cs [hunk 4]
```diff
@@ -183,9 +213,16 @@ private static ActionContext GetActionContext(HttpContext httpContext)
                                     new ActionDescriptor());
         }
 
-        private static HttpContext GetHttpContext()
+        private static HttpContext GetHttpContext(bool addBufferingFeature = false)
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ContentResultTest.cs]: Another nitpick, sorry for not noticing. You might want to just put this inline in your tests that use the feature. That makes it much more apparent what you're doing.

**Comment by rynowak** [comment]
:shipit:

**Comment by pranavkm** [comment]
:shipit:

File: test/Microsoft.AspNet.Mvc.Core.Test/ContentResultTest.cs (hunk 3 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Core.Test/ContentResultTest.cs]: I did not add it explicitly as buffering is not *on* by default and all the remaining tests run without it.

**Pull request merged**